nestkernel/conn_builder_conngen.cpp

```diff
@@ -20,158 +20,140 @@
  *
  */
 
-#include "conngen.h"
+#include "conn_builder_conngen.h"
 
-// Includes from libnestutil:
-#include "logging.h"
+#ifdef HAVE_LIBNEUROSIM
 
 // Includes from nestkernel:
-#include "exceptions.h"
 #include "kernel_manager.h"
-#include "modelrange.h"
 
 // Includes from sli:
-#include "token.h"
+#include "dictutils.h"
 
 namespace nest
 {
 
-/**
- * Low-level function of the ConnectionGenerator interface for
- * connecting populations of neurons using a connection generator with
- * a value set and custom synapse type.
- *
- * \param cg The ConnectionGenerator describing the connectivity
- * \param source_node_ids A NodeCollection specifying the source population
- * \param target_node_ids A NodeCollection specifying the target population
- * \param params_map A Dictionary mapping the labels "weight" and
- *        "delay" to their indices in the value set
- * \param synmodel_name The name of the synapse model to use for the
- *        connections
- */
-void
-cg_connect( ConnectionGeneratorDatum& cg,
-  const NodeCollectionPTR source_node_ids,
-  const NodeCollectionPTR target_node_ids,
-  const DictionaryDatum& params_map,
-  const Name& synmodel_name )
+ConnectionGeneratorBuilder::ConnectionGeneratorBuilder( NodeCollectionPTR sources,
+  NodeCollectionPTR targets,
+  const DictionaryDatum& conn_spec,
+  const std::vector< DictionaryDatum >& syn_specs )
+  : ConnBuilder( sources, targets, conn_spec, syn_specs )
+  , cg_( ConnectionGeneratorDatum() )
+  , params_map_()
 {
-  const Token synmodel = kernel().model_manager.get_synapsedict()->lookup( synmodel_name );
-  if ( synmodel.empty() )
+  updateValue< ConnectionGeneratorDatum >( conn_spec, "cg", cg_ );
+  if ( cg_->arity() != 0 )
   {
-    throw UnknownSynapseType( synmodel_name.toString() );
+    if ( not conn_spec->known( "params_map" ) )
+    {
+      throw BadProperty( "A params_map has to be given if the ConnectionGenerator has values." );
+    }
+
+    updateValue< DictionaryDatum >( conn_spec, "params_map", params_map_ );
+
+    for ( Dictionary::iterator it = params_map_->begin(); it != params_map_->end(); ++it )
+    {
+      it->second.set_access_flag();
+    }
+
+    if ( syn_specs[ 0 ]->known( names::weight ) or syn_specs[ 0 ]->known( names::delay ) )
+    {
+      throw BadProperty( "Properties weight and delay cannot be specified in syn_spec if the "
+			 "ConnectionGenerator has values." );
+    }
   }
-  const index synmodel_id = static_cast< index >( synmodel );
-  DictionaryDatum dummy_params = new Dictionary();
+}
```

Comment:
You should probably also check if `syn_specs.size() > 1` and throw an error if this is the case. Collocation of synapses should not be possible if we use the `conngen` rule.

Reply:
Ah, yes. Now that you say it, I remember that we talked about this! Fixed in 824caee56. Could you maybe provide me with a code snipped that creates collocated synapses, so I can write a test that checks if the error reporting works? Thanks!

Reply:
A simple example would be:
```
node = nest.Create('iaf_psc_alpha', 3)
nest.Connect(node, node, syn_spec=nest.CollocatedSynapses({'weight': -2.}, {'weight': 2.}))
```

Reply:
If you want to test without weight, you could use other parameters than weights, like `synapse_model` and other parameters that can be set for the individual synapse types.
```
node = nest.Create('iaf_psc_alpha', 3)
nest.Connect(node, node, syn_spec=nest.CollocatedSynapses({'synapse_model': 'stdp_synapse', 'alpha': 3.}, {'synapse_model': 'static_synapse'}))
```

Reply:
I've added a test in 500a7412b.